src/lambertw.jl

```diff
@@ -0,0 +1,358 @@
+end
+
+### omega constant ###
+
+const omega_const_ = 0.567143290409783872999968662210355
+# The BigFloat `omega_const_bf_` is set via a literal in the function __init__ to prevent a segfault
+
+# maybe compute higher precision. converges very quickly
+function omega_const(::Type{BigFloat})
+    precision(BigFloat) <= 256 && return omega_const_bf_[]
+    myeps = eps(BigFloat)
+    oc = omega_const_bf_[]
+    for i in 1:100
+        nextoc = (1 + oc) / (1 + exp(oc))
+        abs(oc - nextoc) <= myeps && break
+        oc = nextoc
+    end
+    return oc
+end
+
+"""
+    omega
+    ω
+
+The constant defined by `ω exp(ω) = 1`.
+
+```jldoctest
+julia> ω
+ω = 0.5671432904097...
+
+julia> omega
+ω = 0.5671432904097...
+
+julia> ω * exp(ω)
+1.0
+
+julia> big(omega)
+5.67143290409783872999968662210355549753815787186512508135131079223045793086683e-01
+```
+"""
+const ω = Irrational{:ω}()
+@doc (@doc ω) omega = ω
+
+# The following two lines may be removed when support for v0.6 is dropped
+Base.convert(::Type{AbstractFloat}, o::Irrational{:ω}) = Float64(o)
+Base.convert(::Type{Float16}, o::Irrational{:ω}) = Float16(o)
+Base.convert(::Type{T}, o::Irrational{:ω}) where T <:Number = T(o)
+
+Base.Float64(::Irrational{:ω}) = omega_const_  # FIXME: This is very slow. Why ?
+Base.Float32(::Irrational{:ω}) = Float32(omega_const_)
+Base.Float16(::Irrational{:ω}) = Float16(omega_const_)
+Base.BigFloat(o::Irrational{:ω}) = omega_const(BigFloat)
+
+### Expansion about branch point x = -1/e  ###
+
+# Refer to the paper "On the Lambert W function". In (4.22)
+# coefficients μ₀ through μ₃ are given explicitly. Recursion relations
+# (4.23) and (4.24) for all μ are also given. This code implements the
+# recursion relations.
+
+# (4.23) and (4.24) give zero based coefficients.
+cset(a,i,v) = a[i+1] = v
+cget(a,i) = a[i+1]
+
+# (4.24)
+function compa(k,m,a)
+    sum0 = zero(eltype(m))
+    for j in 2:k-1
+        sum0 += cget(m,j) * cget(m,k+1-j)
+    end
+    cset(a,k,sum0)
+    return sum0
+end
+
+# (4.23)
+function compm(k,m,a)
+    kt = convert(eltype(m),k)
+    mk = (kt-1)/(kt+1) *(cget(m,k-2)/2 + cget(a,k-2)/4) -
+        cget(a,k)/2 - cget(m,k-1)/(kt+1)
+    cset(m,k,mk)
+    return mk
+end
+
+# We plug the known value μ₂ == -1//3 for (4.22) into (4.23) and
+# solve for α₂. We get α₂ = 0.
+# compute array of coefficients μ in (4.22).
+# m[1] is μ₀
+function lamwcoeff(T::DataType, n::Int)
+    # a = @compat Array{T}(undef,n)
+    # m = @compat Array{T}(undef,n)
+    a = zeros(T,n) # We don't need initialization, but Compat is a huge PITA.
+    m = zeros(T,n)
+    cset(a,0,2)  # α₀ literal in paper
+    cset(a,1,-1) # α₁ literal in paper
+    cset(a,2,0)  # α₂ get this by solving (4.23) for alpha_2 with values printed in paper
+    cset(m,0,-1) # μ₀ literal in paper
+    cset(m,1,1)  # μ₁ literal in paper
+    cset(m,2,-1//3) # μ₂ literal in paper, but only in (4.22)
+    for i in 3:n-1  # coeffs are zero indexed
+        compa(i,m,a)
+        compm(i,m,a)
+    end
```

Comment:
I think the code is more readable without the helper functions `cset`, `cget`, `compa`, `compm` (also, if these functions are preserved, the names should be more LambertW-specific)
(I hope I got indexing right)

```suggestion
    a[1] = 2  # α₀ literal in paper
    a[2] = -1 # α₁ literal in paper
    a[3] = 0  # α₂ get this by solving (4.23) for alpha_2 with values printed in paper
    m[1] = -1 # μ₀ literal in paper
    m[2] = 1  # μ₁ literal in paper
    m[3] = -1//3 # μ₂ literal in paper, but only in (4.22)
    @inbounds for i in 4:n
        a[i] = sum(j -> m[j + 1] * m[i + 1 - j], 2:i)
        m[i] = (i-2)/i * (m[i - 2]/2 + a[i - 2]/4) - a[i]/2 - m[i-1]/i
    end
```